### ISSUE TITLE
`mut_mutex_lock` in closures doesn't consider `Fn` bound

### ISSUE THREAD

**Issue by Jasonshengxi** [rust-lang/rust-clippy#13720]
### Summary

When a `&mut Mutex<T>` is given and used in a closure, the lint fails to consider the type constraints on the given closure, and wrongfully suggests `get_mut` which would require a `FnMut` bound instead of `Fn`, creating code that fails to compile.

### Lint Name

mut_mutex_lock

### Reproducer

I tried this code:

```rust
use std::sync::Mutex;

fn _closure(_: impl Fn()) {}
fn _func(mutex: &mut Mutex<()>) {
    _closure(|| {
        let _guard = mutex.lock();
    })
}
```

I saw this happen:

```
warning: calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference
 --> src/lib.rs:6:28
  |
6 |         let _guard = mutex.lock();
  |                            ^^^^ help: change this to: `get_mut`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_mutex_lock
  = note: `#[warn(clippy::mut_mutex_lock)]` on by default
```

I expected to see this happen:
No lint

### Version

```text
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: x86_64-unknown-linux-gnu
release: 1.81.0
LLVM version: 18.1.7
```

**Comment by Jasonshengxi** [comment]
I actually found this while figuring out a different false positive on the same lint, which I minimized to this:
```rust
use std::sync::Mutex;

fn _closure(_: impl FnMut() + 'static) {}
fn _func(mutex: &'static mut Mutex<()>) -> &Mutex<()> {
    _closure(|| {
        let _guard = mutex.lock();
    });
    mutex
}
```
which has a problem that the returned `&Mutex<()>` can't coexist with a `get_mut` mutable borrow which lives for `'static` due to the closure's bounds.

Not sure if this counts as another issue, so I'll leave it here for now.
